xdsl/interpreter.py

```diff
@@ -0,0 +1,273 @@
+    _impls: _InterpreterFunctionImpls = field(
+        default_factory=_InterpreterFunctionImpls)
+    _ctx: InterpreterContext = field(
+        default_factory=lambda: InterpreterContext(name='root'))
+    file: IO[str] | None = field(default=None)
+
+    def get_values(self, values: Iterable[SSAValue]) -> tuple[Any, ...]:
+        """
+        Get values from current environment.
+        """
+        return tuple(self._ctx[value] for value in values)
+
+    def set_values(self, pairs: Iterable[tuple[SSAValue, Any]]):
+        """
+        Set values to current scope.
+        Raises InterpretationError if len(ssa_values) != len(result_values), or if
+        SSA value already has a Python value in the current scope.
+        """
+        for ssa_value, result_value in pairs:
+            self._ctx[ssa_value] = result_value
+
+    def push_scope(self, name: str = 'unknown') -> None:
+        """
+        Create new scope in current environment, with optional custom `name`
+        """
+        self._ctx = InterpreterContext(name, self._ctx)
+
+    def pop_scope(self) -> None:
+        """
+        Discard the current scope, and all the values registered in it. Sets parent scope
+        of current scope to new current scope.
+        Raises InterpretationError if current scope is root scope.
+        """
+        if self._ctx.parent is None:
+            raise InterpretationError('Attempting to pop root env')
+
+        self._ctx = self._ctx.parent
+
+    def register_implementations(self,
+                                 impls: InterpreterFunctions,
+                                 /,
+                                 override: bool = False) -> None:
+        """
+        Register implementations for operations defined in given 
+        `OperationImplementations` object. Raise InterpretationError if an 
+        operation already has an implementation registered, unless override is 
+        set to True.
+        """
+        self._impls.register_from(impls, override=override)
+
+    def run(self, op: Operation):
+        """
+        Fetches the implemetation for the given op, passes it the Python values
+        associated with the SSA operands, and assigns the results to the 
+        operation's results.
+        """
+        inputs = self.get_values(op.operands)
+        results = self._impls.run(self, op, inputs)
+        self._assert(
+            len(op.results) == len(results), 'Incorrect number of results')
+        self.set_values(zip(op.results, results))
+
+    def run_module(self):
+        """
+        Starts execution of `self.module`
+        """
+        self.run(self.module)
+
+    def print(self, *args: Any, **kwargs: Any):
+        """Print to current file."""
+        print(*args, **kwargs, file=self.file)
+
+    def _assert(self, condition: bool, message: str | None = None):
+        "Raise InterpretationError if condition is not satisfied."
+        if not condition:
+            raise InterpretationError(
```

Comment:
I think this should be an assert instead

Reply:
The only place it's used is specifically for when the function returns an incorrect number of parameters, so InterpretationError seems appropriate. The user can always `assert` themselvs if they want an `AssertionError`.

Reply:
Oh okay, then we should rename this function to reflect that!
Something like `raise_error_if` or something?

Reply:
renamed to `interpreter_assert`